### PR TITLE
Remove required columns

### PR DESCRIPTION
**Pull request by niemela**
This removes the "Required?" and "Nullable?" columns from all endpoint specifications.

Technically this makes the minimal compliant implementation of the API significantly smaller, since now simply implementing the `/` and `/access` with the latter returning empty capabilities and endpoints arrays would be compliant. The requirements are instead moved to the CCS requirements document, so for a WF CCS the same functionality is required.

Right now the CCS requirements only contains the definition for on "maximal" access level. We should define all the levels we actually need. Defining them in terms of `/access` should work pretty well.

One issue with the `/access` solution that became obvious while doing this is that it can't contain information about `/scoreboard`, `/event-feed` (and `/webhooks`). I think we should solve that (in a future PR).

### CODE REVIEW

File: Contest_API.md [hunk 1]
```diff
@@ -252,17 +252,19 @@ absolute timestamps.
...
-| hash     | string  | yes       | MD5 hash of the file referenced.
-| mime     | string  | no        | Mime type of resource.
-| width    | integer | depends   | Width of the image. Required for files with mime type image/*.
-| height   | integer | depends   | Height of the image. Required for files with mime type image/*.
+| Name     | Type      | Description
+| -------- | --------- | -----------
+| href     | string ?  | URL where the resource can be found. Relative URLs are relative to the `baseurl`. Must point to a file of intended mime-type. Resource must be accessible using the exact same (possibly none) authentication as the call that returned this data.
+| filename | string    | POSIX compliant filename. Filenames must be unique within the endpoint object where they are used. I.e. an organization can have (multiple) `logo` and `country_flag` file references, they must all have a different filename, but different organizations may have files with the same filename.
+| hash     | string    | MD5 hash of the file referenced.
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:264]: Agreed.

File: Contest_API.md [hunk 2]
```diff
@@ -530,30 +537,18 @@ In the tables below, the columns are:
...
-    `null` value or because it was not implemented.
+    special considerations. Required means that the property must be present 
+    and must not be `null`. Default values specify how missing or `null` 
+    values should be interpreted.
+
+Note that all results returned from endpoints:
+
+  - Must only have `null` values if the type of the property is `<type> ?`.
+  - Must contain all properties specified in the [Access](#access) endpoint
+    with non-`null` values.
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:549]: > Ambiguous

Ah... yes I see it now :).

> I'm assuming you don't mean that, 

Indeed.

> maybe reword as "Must contain all properties specified in the [Access](#access) endpoint that have non-`null` values."

Agreed, that is better.

File: Contest_Control_System_Requirements.md [hunk 3]
```diff
@@ -964,31 +964,101 @@ problem.
...
-| `/judgements`    | `max_run_time`               |
-| `/runs`          | `run_time`                   |
-| `/awards`        | All required elements.       |
+the [Contest API](contest_api).
+
+The CCS must provice access to the Contest API including the 
+`/scoreboard` and `/event-feed` endpoints as well as consistent with 
+(at least) the following return from the `/access` endpoint:
+
+```json
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_Control_System_Requirements.md:973]: Agreed. I tried to avoid changes in meaning since the PR is already so large.

File: Contest_API.md [hunk 4]
```diff
@@ -252,17 +252,19 @@ absolute timestamps.
     with properties as defined below.
   - Arrays (type **`array of <type>`** in the specification) are built-in JSON 
     arrays of some type defined above.
+  - Nullable types (type **`<type> ?`** in the specification) are either a
+    value of a type defined above, or `null`.
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:256]: > * To me 'null' is a valid value for most types, e.g. you can have a null string.

According to the JSON spec a "string" *clearly* can't be null. OTOH, the JSON spec also does not talk about types *at all*. Specifically it does not define a "boolean" but rather the values "true" and "false".

According to JSON Schema, null is its own type, and thus no other type can take the value null. Interestingly (?) it *does* define a boolean type.

> * We've said before that a null value is equivalent to not specifying a property, so to me this is saying that these properties can be null, but isn't saying that others can't be. i.e. it's not really saying anything.

True, but now when we have `/access` there is an important difference... maybe? A value can certainly be missing because it's simply not listed in the access array, but if it *is* and it's not nullable then it can't be null, and therefore it must be there. I.e. a non-nullable value can only be missing if `/access` says so. A nullable value can be missing regardless.

> 'd rather flip this around and specify the required properties, e.g. "Required properties are specified with *, these must exist and be non-null."

I don't *like* it (which is completely subjective) and I think it's not consistent with JSON and JSON Schema (which is at least somewhat objective).

File: Contest_API.md [hunk 5]
```diff
@@ -605,29 +600,28 @@ Provides information on the current contest.
 
 The following endpoints are associated with contest:
 
-| Endpoint         | Mime-type        | Required? | Description                                                                 
-| :--------------- | :--------------- | :-------- | :----------
-| `/contests`      | application/json | yes       | JSON array of all contests with properties as defined in the table below.
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:610]: I'm not so convinced that there are many endpoints that *must* always be present, if by must we mean that you could not have any useful service without them.

These would be my candidates:

- `/` because it tells you which version of the API you have.
- `/contests/<id>/access` because it tells you what other endpoints are availeble.
- `/contests` because otherwise you couldn't know where to get the endpoint above.

The argument for why e.g. `/teams` and `/problems` wouldn't be *always* *strictly* required is that you could use an API for partial configuration of ome other service, and maybe that configuration source doesn't have the teams and problems? OTOH, the counterarguemnt would be that in that case you could just let `/problems` and `/teams` return empty results, so it would not be any additional burden to require them.

In summary, I'm just not sure.

File: Contest_API.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nickygerritsen) [reply Contest_API.md:256]: I think Tim is using java too much, where every object type can also be null 😅 .
I agree that null should be its own type, but I also see the advantage of using `*`. However, I'm fine with the `?` syntax.

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickygerritsen) [reply Contest_API.md:610]: I'd say if you have `/contests` it would be very strange to not also have `/contests/<id>`.

And I kinda agree that there *could* be very strange cases where you don't have `/problems`. I'm not sure if it ever happens in practice and if we are not making it too complex for ourselves by making everything optional though.

File: Contest_Control_System_Requirements.md [hunk 6]
```diff
@@ -964,31 +964,101 @@ problem.
...
-| `/teams`         | `display_name`               |
-| `/teams`         | `organization_id`            |
-| `/teams`         | `group_ids`                  |
-| `/judgements`    | `max_run_time`               |
-| `/runs`          | `run_time`                   |
-| `/awards`        | All required elements.       |
+the [Contest API](contest_api).
+
+The CCS must provice access to the Contest API including the 
+`/scoreboard` and `/event-feed` endpoints as well as consistent with
```
Comment (nickygerritsen) [inline Contest_Control_System_Requirements.md:970]:
```suggestion
`/scoreboard` and `/event-feed` endpoints as well as be consistent with 
```

I think?

File: Contest_Control_System_Requirements.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickygerritsen) [reply Contest_Control_System_Requirements.md:973]: I think the JSON is way more precise. If you are really a CCS considering running for a WF, this JSON should not scare you. If it does, maybe reconsider running at the finals 🤔 🤣 ?

File: Contest_API.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:256]: Yeah, I probably do use Java too much. 😛

I still think what we have is a set of properties that must exist in every response (i.e. `*`), and the rest are optional. The fact that this implies the rest can be null (except booleans) is secondary IMHO and why I'd rather focus on the required properties.

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:610]: I don't love it, but maybe @niemela is right and we should look at this like the recent change to the CAF, where instead of having a bunch of required files we limit this and look at it as 'useful subsets' for certain scenarios.

I think I'd prefer that / and /contests always exist, and when a particular contest exists then /access, /scoreboard (even a dummy/empty response), and /event-feed also exist.

[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:610]: > I'd say if you have `/contests` it would be very strange to not also have `/contests/<id>`.

Yes, that was meant to be implied.

> 'm not sure if it ever happens in practice 

Agreed.

> and if we are not making it too complex for ourselves by making everything optional though.

But, either it never happens, and then it's really not complex. We are not *adding* anything to make things optional, or if it *does* happen, then at least it  wasn't unnecessary, and it's still not very complex?

> I think I'd prefer that / and /contests always exist, and when a particular contest exists then /access, /scoreboard (even a dummy/empty response), and /event-feed also exist.

But scoreboard and event-feed would be two of the things that would be among the more likely/useful to leave out. An API with everything except scoreboard and/or event-feed would certainly be useful, and it would also be *a lot* easier to implement.

[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:610]: Hmm. I guess I grudgingly agree? :)

I'll fall back to / and /contests always existing, and /access exists the moment the contest does.

Access endpoints seems a bit weird for /scoreboard and /event-feed though, maybe these should be capabilities?? We should avoid doing everything in this PR though, I would be fine closing on properties and the basics of which endpoints are mandatory and then handling things like this via another PR or discussion on Monday.

File: Contest_API.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md]: Fixed.

File: Contest_API.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md]: Fixed.

**Comment by eldering** [comment]
> Technically this makes the minimal compliant implementation of the API significantly smaller, since now simply implementing the `/` and `/access` with the latter returning empty capabilities and endpoints arrays would be compliant. The requirements are instead moved to the CCS requirements document, so for a WF CCS the same functionality is required.

I think we should retain some general mimimal set of required endpoints and attributes to implement. For the WFs we can use the CCS specification, but if you're not working in that context, I think we should provide some indication in this specification of what we consider a minimal set of requirements that any client should be able to depend upon.

That can now be specified in terms of the `/access` endpoint, although it might still be helpful to have it listed in each endpoint; not sure about that.

File: Contest_API.md [hunk 7]
```diff
@@ -1604,24 +1603,24 @@ Judgements for submissions in the contest.
...
-| start\_time          | TIME    | yes       | no        | Absolute time when judgement started.
-| start\_contest\_time | RELTIME | yes       | no        | Contest relative time when judgement started.
-| end\_time            | TIME    | yes       | yes       | Absolute time when judgement completed.
-| end\_contest\_time   | RELTIME | yes       | yes       | Contest relative time when judgement completed.
-| max\_run\_time       | number  | no        | yes       | Maximum run time in seconds for any test case. Should be an integer multiple of `0.001`.
+| Name                 | Type      | Description
+| :------------------- | :-------- | :----------
+| id                   | ID        | Identifier of the judgement.
+| submission\_id       | ID        | Identifier of the [ submission](#submissions) judged.
+| judgement\_type\_id  | ID ?      | The [ verdict](#judgement-types) of this judgement.
```
Comment (eldering) [inline Contest_API.md:1617]:
```suggestion
| judgement\_type\_id  | ID ?      | The [ verdict](#judgement-types) of this judgement. Required iff judgement has completed.
```

File: Contest_API.md [hunk 8]
```diff
@@ -1604,24 +1603,24 @@ Judgements for submissions in the contest.
...
+| score                | number    | Score for this judgement. Required iff contest:scoreboard_type is `score`.
+| start\_time          | TIME      | Absolute time when judgement started.
+| start\_contest\_time | RELTIME   | Contest relative time when judgement started.
+| end\_time            | TIME ?    | Absolute time when judgement completed.
+| end\_contest\_time   | RELTIME ? | Contest relative time when judgement completed.
```
Comment (eldering) [inline Contest_API.md:1622]: Should these be required iff `judgement_type_id` is present?

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:610]: > I'd really like to see a "minimal" set of required endpoints and attributes

What minimum set of endpoints do you suggest?

Note that it's not actually a big ask to require an endpoint. All endpoints (except access and state, and access is already required anyway) are sets, which means that the empty set *is* an acceptable implementation. E.g. if awards was required you could trivially implement awards by always returning `[]`, i.e. you would "support" awards, you would just never have any.

File: Contest_Control_System_Requirements.md [hunk 9]
```diff
@@ -964,31 +964,101 @@ problem.
...
-| `/awards`        | All required elements.       |
+the [Contest API](contest_api).
+
+The CCS must provice access to the Contest API including the 
+`/scoreboard` and `/event-feed` endpoints as well as consistent with 
+(at least) the following return from the `/access` endpoint:
+
+```json
+{
+   "capabilities": [],
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_Control_System_Requirements.md:975]: We haven't required that in the past.

File: Contest_Control_System_Requirements.md [hunk 10]
```diff
@@ -964,31 +964,101 @@ problem.
...
+            "max_run_time"
+         ]
+      },
+      { 
+         "type": "runs", 
+         "properties": [
+            "id", "judgement_id", "ordinal", "judgement_type_id", 
+            "time", "contest_time", "run_time"
+         ]
+      },
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_Control_System_Requirements.md:1051]: Maybe, but why?

There is nothing that actually uses it, is there?

File: Contest_API.md [hunk 11]
```diff
@@ -497,14 +499,19 @@ are mentioned below.
 
 ### Types of endpoints
 
-The endpoints can be categorized into 3 groups as follows:
+The endpoints can be categorized into 4 groups as follows:
 
+  - Metadata: `/`, access
```
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:504]: I agree, and I was actually already working on that change, but was on the fence regarding if it should be in the PR or a follow up.

File: Contest_Control_System_Requirements.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (nickygerritsen) [reply Contest_Control_System_Requirements.md:975]: But we clearly have used it for the last few years when John pauses and resumes the contest from his iPad and the CDS forwards this to the primary CCS?

File: Contest_Control_System_Requirements.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (eldering) [reply Contest_Control_System_Requirements.md:1051]: Ok, that's a fair point.

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:610]: I don't think there's any point in requiring endpoints where the expectation is you might always return `[]`. Even for awards, there is no point supporting it unless at a minimum some other tool could PUT and you return that set - i.e. a CCS that can't assign awards but could still be the single source of truth.

From a spec perspective, I think I'm still with my previous comment - the required set should be very small, but instead maybe we define some SHOULDs or logical sets as the CAF examples or CCS spec is doing? e.g. for configuration the CMS (or we) could host a live API with just the registration data; for finals the CCS needs a bigger set?

File: Contest_API.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:256]: A couple days later and I'm still thinking about this as required properties. As a server implementor I want to know what properties I must support and always return from endpoints; as a client I want to know which properties will always be there. I'm happy to be overruled, but I still think we should flip it.

File: Contest_Control_System_Requirements.md [hunk 12]
```diff
@@ -964,31 +964,101 @@ problem.
...
+
+```json
+{
+   "capabilities": [],
+   "endpoints": [
+      { 
+         "type": "contest", 
+         "properties": [
+            "id", "name", "formal_name", "start_time", "duration", 
+            "scoreboard_freeze_duration", "penalty_time"
```
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_Control_System_Requirements.md:981]: No. Most years the CCS hasn't supported pause times, so when the CDS clears the contest start it remembers the current time as a pause, it can be manipulated outside of the CCS, and then when resumed it just sets the start time on the CCS accordingly.

File: Contest_API.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nickygerritsen) [reply Contest_API.md:256]: I have to say reading all the ? didn’t make me very happy either. I know I said I didn’t care but actually reading it changed my opinion a bit

File: Contest_API.md [hunk 13]
```diff
@@ -497,14 +499,19 @@ are mentioned below.
...
   - Configuration: contests, judgement-types, languages, problems,
     groups, organizations, teams, team-members;
   - Live data: state, submissions, judgements, runs, clarifications,
     awards, commentary;
   - Aggregate data: scoreboard, event-feed.
 
+Metadata is data about the API. These are not included in the event feed and
+are always required to be available. The access endpoint specifies which other
+endpoints are offered by the API.
```
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:513]: The Extensibility section should be clear enough that additional un-spec'd properties are allowed.

File: Contest_API.md (hunk 8 above)
Comment (deboer-tim) [reply Contest_API.md:1622]: +1. I think existence of those three should be tied together.

File: Contest_API.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:256]: > As a server implementor I want to know what properties I must support and always return from endpoints; as a client I want to know which properties will always be there. I'm happy to be overruled, but I still think we should flip it.

To me these are two separate questions.

Starting with the second one (i.e. "we should flip if"), I don't think it makes sense at all to say that all data types can be null, but we have a special type operator that makes a type non-nullable. It is much more sane to do it as is in the PR now.

The second question is about what minimum set of properties must always be there. This (arguably) has nothing to do with nullable types. As currently written in this PR the minimum set is empty (which admittedly isn't so useful for clients), but you are told exactly what properties you will get in the access endpoint. I'm on the fence regarding increasing the minimum set to something actually useful, but if we do I would suggest specifying it in the access endpoint.

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:610]: > I don't think there's any point in requiring endpoints where the expectation is you might always return `[]`.

Right, but that's true for pretty much *all* endpoints. Which kinda was my point?

Although the point does cut both ways, i.e.:
- It's no big hurdle to be forced to implement (say) clarifications, since the server could just not have any and always return `[]`. Thus there is no reason to *not* require it.
- There is no practical difference from getting a 404 from a non-implemented endpoint compared to getting a `[]` for a unused (but technically implemented) endpoint. Except in the latter case `/access` would still list the unused endpoint and would this be slightly less useful. Thus there is no reason *to* require it.

...

Actually after writing down the argument in either direction I'm leaning a but for towards *not* requiring more endpoints.

File: Contest_Control_System_Requirements.md (hunk 6 above)
Comment (niemela) [reply Contest_Control_System_Requirements.md:970]: Fixed.

File: Contest_Control_System_Requirements.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_Control_System_Requirements.md:975]: Slippery slope exemplified :)

File: Contest_API.md (hunk 7 above)
Comment (niemela) [reply Contest_API.md:1617]: Fixed.

File: Contest_API.md (hunk 8 above)
Comment (niemela) [reply Contest_API.md:1622]: Fixed.

File: Contest_API.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (eldering) [reply Contest_API.md:513]: I meant it in the sense that all attributes currently specified in the `/` and `access` endpoints are required, but in the future we could specify optional attributes there. Do we already want to accomodate for that in the text and tables now?

[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:513]: The fact that `/access` specifies endpoints by reference to a table in the spec means that it's not (really) possible to add extensions in the form of new endpoints. Do we want to fix that?

[the remark this answers is not in the document]
Comment (eldering) [reply Contest_API.md:513]: > The fact that `/access` specifies endpoints by reference to a table in the spec means that it's not (really) possible to add extensions in the form of new endpoints. Do we want to fix that?

Yes, I think it would be good to somehow mention that you can provide unspecified endpoints or properties, and then you should also include these in the `access` endpoint. That's how I interpret the extensibility section.

But there's also the meta-point, that the `access` endpoint doesn't specify what properties are available in the meta-data endpoints `/` or `access` themselves. I guess we don't have to tackle that in this PR, but would be good to put on the list with scoreboard and event-feed endpoints to someone add to `access`.

[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:513]: > Yes, I think it would be good to somehow mention that you can provide unspecified endpoints or properties, and then you should also include these in the `access` endpoint.

Right, but you can't actually do that. The values for the "type" property refers to the notification format table for valid values. Any unspecified endpoint that we want to include will, by definition, not be in that list, so it can't be added to the `access` endpoint.

> But there's also the meta-point, that the `access` endpoint doesn't specify what properties are available in the meta-data endpoints `/` or `access` themselves.

I don't think that's a big problem. Both of those endpoints are specified to be always required, and at least for now all properties must always be there.

File: Contest_API.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md]: Fixed.

File: Contest_API.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:512]: I made it more clear that all properties in `/` and `/access` are required.

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (eldering) [reply Contest_API.md:610]: Ok, so I'd propose the following as a "minimal" useful set of endpoints and properties a server should implement:
```json
{
   "capabilities": [],
   "endpoints": [
      {
         "type": "contest",
         "properties": [
            "id", "name", "start_time", "duration"
         ]
      },
      {
         "type": "judgement-types",
         "properties": [
            "id", "name", "solved"
         ]
      },
      {
         "type": "languages",
         "properties": [
            "id", "name"
         ]
      },
      {
         "type": "problems",
         "properties": [
            "id", "name"
         ]
      },
      {
         "type": "teams",
         "properties": [
            "id", "name"
         ]
      },
      {
         "type": "submissions",
         "properties": [
            "id", "language_id", "problem_id", "team_id", "time"
         ]
      },
      {
         "type": "judgements",
         "properties": [
            "id", "submission_id", "judgement_type_id"
         ]
      }
   ]
}
```
I think that for any reasonably normal contest, this is pretty much the bare minimum that allows a client to have the basic information do some generic useful things.

**Comment by eldering** [comment]
> I think we should retain some general mimimal set of required endpoints and attributes to implement. For the WFs we can use the CCS specification, but if you're not working in that context, I think we should provide some indication in this specification of what we consider a minimal set of requirements that any client should be able to depend upon.
> 
> That can now be specified in terms of the `/access` endpoint, although it might still be helpful to have it listed in each endpoint; not sure about that.

Closing this comment in favor of https://github.com/icpc/ccs-specs/pull/86#discussion_r830682326

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:610]: I think Fredrik's point (a few ago...) was what if (e.g.) someone wanted to host the CMS data on a server instead of a file system? With the list above this would be invalid, or they'd have to fake some empty judgement types, languages, and submissions.

Like the CAF, I am wondering if the API spec really needs all of these required, or if your list above is more of a minimal CCS implementation of the spec.

File: Contest_Control_System_Requirements.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_Control_System_Requirements.md:975]: IMHO this is the difference between 'does a CCS need to implement this' (no), and 'would John require this for a CCS to be primary at WF' (yes).

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:610]: > ```json
>          "type": "contest",
>          "properties": [
>             "id", "name", "start_time", "duration"
> ```

If we don't include `penalty_time` it needs a default value. (0 would be natural).

> ```json
>          "type": "languages",
> ```

I don't think `languages` is needed. Specifically in a "submit the output instead of the code" kind of contest.

> ```json
>          "type": "submissions",
>          "properties": [
>             "id", "language_id", "problem_id", "team_id", "time"
> ```

If we drop `languages` above we must also drop `language_id` here.

I agree that this a a good very minimal contest setup. See @deboer-tim post above though (which I agree with)

[the remark this answers is not in the document]
Comment (eldering) [reply Contest_API.md:610]: Ok, but in response to @deboer-tim's comment: I think of this explicitly as a "should" implement for CCSs, and don't consider this a requirement e.g. for a contest archive.

And I agree that language could indeed be dropped. Although I'd say that it would almost always be available, I think it is not _that_ useful for a typical client, so leaving languages out is ok to me.

Indeed, `penalty_time = 0` seems a reasonable default if nothing is included.

File: Contest_API.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:256]: Yes, you're right. These often overlap but are independent, and also how we ended up with two columns before. :-/  In my head I was assuming that was the required column we were effectively retaining, but you've really been doing the nullable.

I do think it's important to keep some representation of required/minimum. It's now impossible to tell which clarification properties I must implement vs which are optional, or whether I need to implement contest start_time (b/c hey, it's nullable). I don't want to bring back a column, but IMHO it is much better to do this inside each endpoint (i.e. in the 'type' definition) vs merged into a common access endpoint.

File: Contest_Control_System_Requirements.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_Control_System_Requirements.md:975]: I suggest discussing this on the next call.

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (niemela) [reply Contest_API.md:610]: I added a default for `penalty_time`.

File: Contest_API.md [hunk 14]
```diff
@@ -252,17 +252,19 @@ absolute timestamps.
...
-| width    | integer | depends   | Width of the image. Required for files with mime type image/*.
-| height   | integer | depends   | Height of the image. Required for files with mime type image/*.
+| Name     | Type      | Description
+| -------- | --------- | -----------
+| href     | string ?  | URL where the resource can be found. Relative URLs are relative to the `baseurl`. Must point to a file of intended mime-type. Resource must be accessible using the exact same (possibly none) authentication as the call that returned this data.
+| filename | string    | POSIX compliant filename. Filenames must be unique within the endpoint object where they are used. I.e. an organization can have (multiple) `logo` and `country_flag` file references, they must all have a different filename, but different organizations may have files with the same filename.
+| hash     | string ?  | MD5 hash of the file referenced.
+| mime     | string    | Mime type of resource.
+| width    | integer ? | Width of the image. Required for files with mime type image/*.
+| height   | integer ? | Height of the image. Required for files with mime type image/*.
```
Comment (deboer-tim) [inline Contest_API.md:267]: I just realized that this means width and height can be either an integer or 'null', and I don't like that. :) In the case of integers or booleans IMHO it's preferable to leave the value out and there is a default specified ('unknown' in this case).

File: Contest_API.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (deboer-tim) [reply Contest_API.md:610]: Had a chat with Fredrik yesterday, and I'm more comfortable with not specifying required properties in each endpoint. I do think we need to document this somewhere for different uses though. The CAF does this in the examples section; now we do it for the API in the CCS API, but don't cover the additional CCS requirements for finals or other cases. I think we should go ahead and merge this soon (it's now blocking other cases), and talk on the next call about how to cover this.

**Review by eldering — APPROVED** [review]
I'm ok with merging as is, although there are some things we should fix in next iterations.

**Review by deboer-tim — APPROVED** [review]
